### PR TITLE
ARROW-593 [C++]: Rename ReadableFileInterface to RandomAccessFile

### PR DESCRIPTION
**Pull request by JohanMabille**

### CODE REVIEW

File: cpp/src/arrow/io/interfaces.h [hunk 1]
```diff
@@ -129,10 +129,10 @@ class ARROW_EXPORT WriteableFileInterface : public OutputStream, public Seekable
   WriteableFileInterface() { set_mode(FileMode::READ); }
 };
 
-class ARROW_EXPORT ReadWriteFileInterface : public ReadableFileInterface,
+class ARROW_EXPORT ReadWriteFileInterface : public RandomAccessFile,
                                             public WriteableFileInterface {
  protected:
-  ReadWriteFileInterface() { ReadableFileInterface::set_mode(FileMode::READWRITE); }
+  ReadWriteFileInterface() { RandomAccessFile::set_mode(FileMode::READWRITE); }
 };
```
Comment (wesm) [inline cpp/src/arrow/io/interfaces.h]: Would you mind adding a `using ReadableFileInterface = RandomAccessFile;` to ease matters for thirdparty users (this class is used in parquet-cpp, for example)

**Review by wesm — COMMENTED** [review]
Cool, thank you for doing this. This brings this concept more semantically in line with other C++ libraries in the ecosystem like TensorFlow

File: cpp/src/arrow/io/interfaces.h (hunk 1 above)
Comment (JohanMabille) [inline cpp/src/arrow/io/interfaces.h]: Yes, I missed the backward compatibility.

**Review by wesm — APPROVED** [review]
+1, thank you!

**Comment by JohanMabille** [comment]
> This brings this concept more semantically in line with other C++ libraries in the ecosystem like TensorFlow

I agree, it also makes it more consistent with other parts of the standard library, which provides the RandomAccessIterator concept and iterators on streams. So if you plan to provide an iterator interface a la STL for the various interfaces in Arrow, this renaming will completely make sense.

**Pull request closed without merging**